components/bigqueryDS.js

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyvaright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const { Datasource } = require("./datasource.js");
+const _ = require("lodash");
+const exec = require("await-exec");
+
+function safeJSONParse(str) {
+  try {
+    return JSON.parse(str);
+  } catch (err) {
+    console.warn("failed parsing", str);
+    return { err: "failed to parse", bad_str: str };
+  }
+}
+
+function sanitizeTimestamp(str) {
+  // new Date("2022-04-14 09:24:34") generates  'Invalid Date' on safari.  Node seems to be happy ... so fix it up
+  const date = new Date(str);
+  return date.toISOString();
+}
+
+function normalize(entry) {
+  const normalizedEntry = {
+    insertId: entry.insertId,
+    logName: entry.logName,
+    receiveTimestamp: sanitizeTimestamp(entry.receiveTimestamp),
+    timestamp: sanitizeTimestamp(entry.timestamp),
+    jsonPayload: {
+      errorResponse: entry.jsonPayload_errorResponse,
+      request: safeJSONParse(entry.jsonPayload_request),
+      response: safeJSONParse(entry.jsonPayload_response),
+      "@type": entry.jsonPayload_type,
+    },
+    labels: safeJSONParse(entry.labels),
+    resource: {
+      labels: safeJSONParse(entry.resource_labels),
+    },
+  };
+  return normalizedEntry;
+}
+
+class Bigquery extends Datasource {
+  constructor(argv) {
+    super(argv);
+    this.dataset = argv.bigquery;
+    //this.timerangeClause = '
+  }
+
+  async bq(query) {
+    const cmd = `bq query --max_rows=1000 --format=json --nouse_legacy_sql '${query}'`;
+    console.log("Running bq command", cmd);
+    const ret = await exec(cmd, { maxBuffer: 1024 * 1024 * 30 });
+    //console.log("gots output", ret);
+    return safeJSONParse(ret.stdout);
+  }
+  /*
+   * Extract trip_ids from vehicle logs, and query again to
+   * get the createTrip/update trip calls that may not be labeled
+   * with a vehicle
+   */
+  async fetchTripLogsForVehicle(vehicle_id, vehicleLogs) {
+    console.log("Loading trip logs for vehicle id", vehicle_id);
+    const trip_ids = _(vehicleLogs)
+      .map((x) => _.split(_.get(x, "labels.trip_id"), ","))
+      .flatten()
+      .uniq()
+      .compact()
+      .value();
+    const tripPromises = trip_ids.map(async (trip_id) => {
+      const createLogs = await this.bq(
+        `select * from \`${this.dataset}.create_trip\` where json_value(labels.trip_id)="${trip_id}" order by timestamp desc limit 10`
```

Comment:
Consider being more defensive w/ https://cloud.google.com/bigquery/docs/parameterized-queries#node.js

Reply:
mostly addressed with switch to bigquery node library